### DISCUSSION TITLE
Cosmos as FPGA hardware with Hastlayer

### DISCUSSION THREAD

**Discussion by Piedone** [CosmosOS/Cosmos#2838] [Ideas]
This is just a random idea, nothing concrete, to throw in to see what you all think.

We have [Hastlayer](https://github.com/Lombiq/Hastlayer-SDK) that can turn .NET assemblies into FPGA hardware. The logic in the end will fully be implemented as hardware.

The goal of Hastlayer is to provide hardware acceleration for highly parallelized, CPU/memory-bound .NET algorithms, by offloading such computations to an FPGA accelerator card. However, while that's not the primary use case we aim for, the resulting hardware implementation can also run on its own, if it's embedded into a hardware design that satisfies its very simple interface.

While Hastlayer [has its limitations](https://github.com/Lombiq/Hastlayer-SDK/blob/dev/Docs/WorkingWithHastlayer.md#writing-hastlayer-compatible-net-code), and I'm sure it'd take a lot of work to actually do it, but in principle, it could be used to turn Cosmos into a piece of hardware too.

Do you think this would make sense, would it be an exciting thing to try?

**Comment by zarlo** [comment]
the biggest issues would likely be that cosmos injects code replacements at compile time after the IL is compiled it injects both new IL and ASM so it might be hard to use the output from a cosmos project

but saying that you can override just about every thing we inject with what ever you want you could even stop them from been in injected but if you so that a lot of methods will just throw an exception

**Reply by Piedone** [reply]
That would indeed be a problem. Injected IL not so much, since Hastlayer processes the finished DLL files (it doesn't hook into the build), but ASM won't work.

**Reply by zarlo** [reply]
when we inject the IL its after the dll is made the injected IL only ends up in the .asm file and .bin file

**Reply by Piedone** [reply]
I see, that's incompatible with Hastlayer, then.

**Reply by zarlo** [reply]
we "patch" the dotnet standard library at build time in memory, but we might be able to dump file with the patches to disk

**Reply by Piedone** [reply]
That would be necessary. However, does that mean that you also IL weave .NET itself?

The idea comes from @plasticalligator [here](https://github.com/Lombiq/Hastlayer-SDK/discussions/122) BTW.

**Reply by Piedone** [reply]
FYI some interesting food for thought [here](https://github.com/Lombiq/Hastlayer-SDK/discussions/122#discussioncomment-7804774) about using Fully Homomorphic Encryption for inter-process communication.
